File: packages/contratos/test/stock.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';

import {
  CASOS_DE_MOVIMIENTO,
  CODIGOS_DE_RECHAZO,
  MOTIVOS_DE_AJUSTE,
  TOPE_DE_STOCK,
  aplicarOperacion,
  parsearPedidoDeMovimiento,
  type PedidoDeMovimiento,
} from '../src/stock.ts';

type OperacionDeStock = PedidoDeMovimiento["operacion"];

const ID = '3f2b9c1e-7a4d-4e8f-9b21-5c6d7e8f9a0b';

const pedido = (operacion: unknown, cambios: Record<string, unknown> = {}) => ({
  productoId: 'trumpeter-malbec',
  idMovimiento: ID,
  operacion,
  ...cambios,
});

const reponer = (cantidad: unknown) => pedido({ tipo: 'reponer', cantidad });
const corregir = (visto: unknown, valor: unknown, motivo: unknown = 'conteo') =>
  pedido({ tipo: 'corregir', visto, valor, motivo });

// ---------------------------------------------------------------- el parser

test('control positivo: una reposicion y una correccion validas pasan', () => {
  const r = parsearPedidoDeMovimiento(reponer(6));
  assert.ok(r.ok);
  assert.deepEqual(r.valor.operacion, { tipo: 'reponer', cantidad: 6 });

  const c = parsearPedidoDeMovimiento(corregir(10, 7, 'rotura'));
  assert.ok(c.ok);
  assert.deepEqual(c.valor.operacion, { tipo: 'corregir', visto: 10, valor: 7, motivo: 'rotura' });
});

test('reponer: cero, negativos, fracciones, textos y NaN se rechazan', () => {
  for (const cantidad of [0, -1, 1.5, '6', null, undefined, NaN, Infinity]) {
    assert.equal(parsearPedidoDeMovimiento(reponer(cantidad)).ok, false, `cantidad ${String(cantidad)}`);
  }
});

test('reponer: el tope de una sola reposicion es TOPE_DE_STOCK, ni una mas', () => {
  assert.ok(parsearPedidoDeMovimiento(reponer(TOPE_DE_STOCK)).ok);
  assert.equal(parsearPedidoDeMovimiento(reponer(TOPE_DE_STOCK + 1)).ok, false);
});

test('corregir: el valor puede ser 0 (se vacio el deposito) pero no negativo ni pasado el tope', () => {
  assert.ok(parsearPedidoDeMovimiento(corregir(5, 0)).ok);
  assert.equal(parsearPedidoDeMovimiento(corregir(5, -1)).ok, false);
  assert.ok(parsearPedidoDeMovimiento(corregir(5, TOPE_DE_STOCK)).ok);
  assert.equal(parsearPedidoDeMovimiento(corregir(5, TOPE_DE_STOCK + 1)).ok, false);
});

test('corregir: `visto` sin tope propio, para poder bajar un dato anterior a la baranda', () => {
  assert.ok(parsearPedidoDeMovimiento(corregir(TOPE_DE_STOCK + 900, 4000)).ok);
  assert.equal(parsearPedidoDeMovimiento(corregir(-1, 4)).ok, false);
  assert.equal(parsearPedidoDeMovimiento(corregir(2.5, 4)).ok, false);
});

test('el cero negativo se rechaza en cada campo: Firestore lo guardaria como double', () => {
  // `Number.isInteger(-0)` y `-0 >= 0` son true: sin `Object.is` pasaria, se
  // guardaria `doubleValue: 0` y `d.stock is int` de las reglas fallaria.
  // Control positivo: el 0 comun de cada campo si pasa.
  assert.ok(parsearPedidoDeMovimiento(corregir(5, 0)).ok);
  assert.ok(parsearPedidoDeMovimiento(corregir(0, 5)).ok);
  assert.equal(parsearPedidoDeMovimiento(corregir(5, -0)).ok, false, 'valor -0');
  assert.equal(parsearPedidoDeMovimiento(corregir(-0, 5)).ok, false, 'visto -0');
  assert.equal(parsearPedidoDeMovimiento(reponer(-0)).ok, false, 'cantidad -0');
});

test('corregir al mismo valor visto no es una correccion', () => {
  assert.equal(parsearPedidoDeMovimiento(corregir(10, 10)).ok, false);
});

test('corregir: el motivo es de la lista cerrada, y cada uno de la lista pasa', () => {
  for (const motivo of MOTIVOS_DE_AJUSTE) {
    assert.ok(parsearPedidoDeMovimiento(corregir(10, 7, motivo)).ok, motivo);
  }
  // Sin pasar por el helper: su valor por defecto convertiria `undefined` en
  // 'conteo' y este caso -el motivo AUSENTE- pasaria sin probar nada.
  for (const motivo of ['porque-si', '', null, 3, undefined]) {
    const operacion = { tipo: 'corregir', visto: 10, valor: 7, motivo };
    assert.equal(parsearPedidoDeMovimiento(pedido(operacion)).ok, false, String(motivo));
  }
  const sinMotivo = { tipo: 'corregir', visto: 10, valor: 7 };
  assert.equal(parsearPedidoDeMovimiento(pedido(sinMotivo)).ok, false, 'sin la clave motivo');
});

test('el pedido se cierra: una clave de mas, en cualquier nivel, lo rechaza', () => {
  assert.equal(parsearPedidoDeMovimiento(pedido({ tipo: 'reponer', cantidad: 6 }, { stock: 999 })).ok, false);
  assert.equal(parsearPedidoDeMovimiento(pedido({ tipo: 'reponer', cantidad: 6, despues: 999 })).ok, false);
  assert.equal(
    parsearPedidoDeMovimiento(pedido({ tipo: 'corregir', visto: 1, valor: 2, motivo: 'otro', por: 'yo' })).ok,
    false,
  );
});

test('productoId: sin barra, sin punto, sin ids reservados de Firestore', () => {
  for (const productoId of ['', 'a/b', '../x', 'a.b', '__reservado__', 12, null, 'x'.repeat(129)]) {
    assert.equal(
      parsearPedidoDeMovimiento(pedido({ tipo: 'reponer', cantidad: 1 }, { productoId })).ok,
      false,
      String(productoId),
    );
  }
  assert.ok(
    parsearPedidoDeMovimiento(pedido({ tipo: 'reponer', cantidad: 1 }, { productoId: 'muestra-catena-zapata' })).ok,
  );
});

test('idMovimiento: entre 16 y 64 caracteres seguros', () => {
  for (const idMovimiento of ['', 'corto', 'x'.repeat(65), 'con espacios y demas cosas!!', 7, null]) {
    assert.equal(
      parsearPedidoDeMovimiento(pedido({ tipo: 'reponer', cantidad: 1 }, { idMovimiento })).ok,
      false,
      String(idMovimiento),
    );
  }
});

test('lo que no es un objeto no es un pedido', () => {
  for (const x of [null, undefined, 3, 'reponer', [], [reponer(1)]]) {
    assert.equal(parsearPedidoDeMovimiento(x).ok, false);
  }
  assert.equal(parsearPedidoDeMovimiento(pedido('reponer')).ok, false);
  assert.equal(parsearPedidoDeMovimiento(pedido({ tipo: 'vaciar' })).ok, false);
});

test('el orden de las claves de la operacion parseada es fijo, sea cual sea el de entrada', () => {
  // La callable compara un reintento contra lo guardado por JSON.stringify:
  // solo sirve si el parser arma siempre el mismo orden.
  const a = parsearPedidoDeMovimiento(pedido({ tipo: 'corregir', visto: 3, valor: 9, motivo: 'conteo' }));
  const b = parsearPedidoDeMovimiento(pedido({ motivo: 'conteo', valor: 9, tipo: 'corregir', visto: 3 }));
  assert.ok(a.ok && b.ok);
  assert.equal(JSON.stringify(a.valor.operacion), JSON.stringify(b.valor.operacion));
});

// ------------------------------------------------------------ la cuenta

const rep = (cantidad: number): OperacionDeStock => ({ tipo: 'reponer', cantidad });
const cor = (visto: number, valor: number): OperacionDeStock => ({
  tipo: 'corregir',
  visto,
  valor,
  motivo: 'conteo',
});

test('reponer suma sobre el stock actual, y hasta el tope justo', () => {
  assert.deepEqual(aplicarOperacion(4, rep(6)), { ok: true, despues: 10 });
  assert.deepEqual(aplicarOperacion(0, rep(1)), { ok: true, despues: 1 });
  assert.deepEqual(aplicarOperacion(TOPE_DE_STOCK - 10, rep(10)), { ok: true, despues: TOPE_DE_STOCK });
});

test('reponer pasado el tope se rechaza, aunque cada numero por separado sea valido', () => {
  assert.deepEqual(aplicarOperacion(TOPE_DE_STOCK - 10, rep(11)), { ok: false, codigo: 'pasa-el-tope' });
  assert.deepEqual(aplicarOperacion(TOPE_DE_STOCK, rep(1)), { ok: false, codigo: 'pasa-el-tope' });
});

test('corregir fija el valor cuando lo visto coincide', () => {
  assert.deepEqual(aplicarOperacion(10, cor(10, 7)), { ok: true, despues: 7 });
  assert.deepEqual(aplicarOperacion(5, cor(5, 0)), { ok: true, despues: 0 });
});

test('corregir sobre un stock que cambio se rechaza: no pisa las ventas del medio', () => {
  // El operador vio 10, conto 7, y mientras tanto se vendieron dos: el stock
  // es 8.  Fijar 7 borraria una de las dos ventas.
  assert.deepEqual(aplicarOperacion(8, cor(10, 7)), { ok: false, codigo: 'cambio-el-stock' });
  // Y tambien si subio: entro una reposicion mientras contaba.
  assert.deepEqual(aplicarOperacion(16, cor(10, 7)), { ok: false, codigo: 'cambio-el-stock' });
});

test('reponer no mira lo visto: es conmutativa, dos reposiciones seguidas suman las dos', () => {
  const primera = aplicarOperacion(4, rep(6));
  assert.ok(primera.ok);
  assert.deepEqual(aplicarOperacion(primera.despues, rep(6)), { ok: true, despues: 16 });
});

// ------------------------------------------------------------- las fixtures

test('las fixtures cubren las cuatro salidas posibles', () => {
  const salidas = new Set<string>();
  for (const c of CASOS_DE_MOVIMIENTO) {
    const p = parsearPedidoDeMovimiento(pedido(c.operacion));
    if (!p.ok) {
      salidas.add('invalido');
      continue;
    }
    const a = aplicarOperacion(c.stock, p.valor.operacion);
    salidas.add(a.ok ? 'aplica' : a.codigo);
  }
  // Si una de las cuatro clases faltara, el espejo en Dart no la probaria.
  assert.deepEqual([...salidas].sort(), ['aplica', 'cambio-el-stock', 'invalido', 'pasa-el-tope']);
  for (const codigo of CODIGOS_DE_RECHAZO) assert.ok(salidas.has(codigo), codigo);
});

test('los `porque` de las fixtures no se repiten', () => {
  const porques = CASOS_DE_MOVIMIENTO.map((c) => c.porque);
  assert.equal(new Set(porques).size, porques.length);
});

File: apps/admin/test/core/contratos/stock_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:admin/core/contratos/stock.dart';
import 'package:test/test.dart';

/// El espejo de `packages/contratos/src/stock.ts`. Compara CONTRA
/// `generated/contratos.json`, no contra si mismo: los casos y sus salidas los
/// CALCULO el TypeScript de hoy.
///
/// Usa package:test y NO flutter_test A PROPOSITO: `flutter test` esta
/// DENEGADO en esta maquina.
void main() {
  late Map<String, dynamic> stock;

  setUpAll(() {
    final archivo = File('../../packages/contratos/generated/contratos.json');
    final contrato =
        json.decode(archivo.readAsStringSync()) as Map<String, dynamic>;
    stock = contrato['stock'] as Map<String, dynamic>;
  });

  test('topeDeStock coincide con el contrato', () {
    expect(topeDeStock, equals(stock['topeDeStock']));
  });

  test('MotivoDeAjuste cubre exactamente los motivos del contrato', () {
    expect(
      MotivoDeAjuste.values.map((m) => m.clave).toList(),
      equals(List<String>.from(stock['motivosDeAjuste'] as List)),
    );
  });

  test('CodigoDeRechazo cubre exactamente los codigos del contrato', () {
    expect(
      CodigoDeRechazo.values.map((c) => c.clave).toList(),
      equals(List<String>.from(stock['codigosDeRechazo'] as List)),
    );
  });

  test(
    'MotivoDeAjuste.desde: una clave desconocida da null, no una por defecto',
    () {
      expect(MotivoDeAjuste.desde('conteo'), MotivoDeAjuste.conteo);
      expect(MotivoDeAjuste.desde('porque-si'), isNull);
      expect(MotivoDeAjuste.desde(null), isNull);
      expect(MotivoDeAjuste.desde(3), isNull);
    },
  );

  group('las fixtures del contrato', () {
    late List casos;

    setUpAll(() {
      casos = stock['casos'] as List;
    });

    test('el contrato trae fixtures', () {
      // Una lista vacia haria pasar el resto sin comparar nada.
      expect(casos, isNotEmpty);
    });

    test('las fixtures tienen las cuatro salidas: aplica, las dos de rechazo y '
        'la invalida', () {
      final salidas = <String>{};
      for (final c in casos.cast<Map<String, dynamic>>()) {
        if (c['valido'] != true) {
          salidas.add('invalido');
        } else {
          final r = c['resultado'] as Map<String, dynamic>;
          salidas.add(r['ok'] == true ? 'aplica' : r['codigo'] as String);
        }
      }
      expect(
        salidas,
        equals({'aplica', 'cambio-el-stock', 'pasa-el-tope', 'invalido'}),
      );
    });

    test('cada caso: la validez y la cuenta dan lo que dijo TypeScript', () {
      var comparados = 0;
      var invalidosPorTipo = 0;

      for (final c in casos.cast<Map<String, dynamic>>()) {
        final porque = c['porque'] as String;
        final operacion = _aDart(c['operacion']);

        // Lo que Dart ni siquiera puede escribir -un texto donde va un
        // numero, una fraccion, un motivo que no existe- es invalido en
        // TypeScript tambien. Es el tipo quien lo frena aca.
        if (operacion == null) {
          expect(c['valido'], isFalse, reason: porque);
          invalidosPorTipo++;
          continue;
        }

        comparados++;
        expect(operacion.esValida, equals(c['valido']), reason: porque);
        if (c['valido'] != true) continue;

        final esperado = c['resultado'] as Map<String, dynamic>;
        final obtenido = aplicarOperacion(c['stock'] as int, operacion);
        if (esperado['ok'] == true) {
          expect(obtenido, isA<Aplica>(), reason: porque);
          expect(
            (obtenido as Aplica).despues,
            esperado['despues'],
            reason: porque,
          );
        } else {
          expect(obtenido, isA<Rechaza>(), reason: porque);
          expect(
            (obtenido as Rechaza).codigo.clave,
            esperado['codigo'],
            reason: porque,
          );
        }
      }

      // Si `_aDart` devolviera null para todo, el bucle "pasaria" sin
      // comparar nada: los dos numeros tienen que ser distintos de cero.
      expect(comparados, greaterThan(8));
      expect(invalidosPorTipo, greaterThan(2));
    });
  });

  group('aJson: lo que viaja a la callable', () {
    test('reponer lleva {tipo, cantidad} y nada mas', () {
      expect(
        const Reponer(6).aJson,
        equals({'tipo': 'reponer', 'cantidad': 6}),
      );
    });

    test('corregir lleva {tipo, visto, valor, motivo} y nada mas', () {
      expect(
        const Corregir(
          visto: 10,
          valor: 7,
          motivo: MotivoDeAjuste.rotura,
        ).aJson,
        equals({
          'tipo': 'corregir',
          'visto': 10,
          'valor': 7,
          'motivo': 'rotura',
        }),
      );
    });
  });
}

/// De lo que trae el JSON a lo que Dart puede escribir, o `null` si el tipo
/// mismo lo impide.
OperacionDeStock? _aDart(Object? json) {
  if (json is! Map) return null;
  final tipo = json['tipo'];
  if (tipo == 'reponer') {
    final cantidad = json['cantidad'];
    return cantidad is int ? Reponer(cantidad) : null;
  }
  if (tipo == 'corregir') {
    final visto = json['visto'];
    final valor = json['valor'];
    final motivo = MotivoDeAjuste.desde(json['motivo']);
    if (visto is int && valor is int && motivo != null) {
      return Corregir(visto: visto, valor: valor, motivo: motivo);
    }
  }
  return null;
}
